[dynamo] [guard] Add caching for inside torch.compile.disable function to avoid unnecessary recompilation.

thenumberouscode opened the pull request:
inside torch.compile.disable function always triggers recompilation. because a user inside function decorated with torch._dynamo.disable would be used as an argument in the resume_in_xx function. In the current implementation,  it will always be a new object, resulting in the ID_MATCH guard always failing and triggering recompilation. 

Fixes https://github.com/pytorch/pytorch/issues/157399
@xmfan 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela @mlazos

thenumberouscode commented:
@pytorchbot label "release notes: dynamo"

thenumberouscode commented:
@zou3519 Could you please take a look at my code? I would really appreciate your feedback

jmaczan commented on torch/_dynamo/variables/functions.py:
```diff
@@ -1387,7 +1387,9 @@ def as_python_constant(self):
     def create_with_source(cls, value, source):
-        if not is_wrapper_or_member_descriptor(value):
+        if inspect.getattr_static(value, "_torchdynamo_disable", False):
+            install_guard(source.make_guard(GuardBuilder.TYPE_MATCH))
```
I'm new to dynamo so forgive silly question, but could we add a new rule to trace_rules.manual_torch_name_rule_map instead?

```py
"torch._dynamo.disable": SkipFunctionVariable,
```

It wouldn't introduce additional code, but I didn't test it beside a happy path. @anijain2305 @thenumberouscode

thenumberouscode replied:
> Maybe install the FUNCTION_MATCH on the `value._torchdynamo_disable`. Something along the lines of
> 
> ```
> 
>     fn_obj = value._torchdynamo_disable
>     install_guard(AttrSource(source, "_torchdynamo_disable")).make_guard(GuardBuilder.FUNCTION_MATCH))
> ```

Your code indeed offers a more granular solution. I'm just waiting for the results of all the workflows

jmaczan replied:
I just gave it a quick test to see if the bug occurs and it doesn’t, but didn’t test it further

mlazos replied:
@anijain2305 I don't think this is sound - `_torchdynamo_disable` is only ever `True` right? At the root of the issue is that the ID of a local nested function is changing on each call, I don't think we can guarantee that it's semantically the same function though (which is why we usually check function ID). We could have any number of closures inside the function and pass them as arguments to the resume function based on different conditions and we wouldn't recompile the resume function even though we should.

thenumberouscode replied:
@mlazos I’m a bit confused about something. To me, a function that doesn’t need to be compiled means we don’t have to worry about its implementation, arguments, or any other details; we just pass it to Python’s default interpreter.

thenumberouscode replied:
Cool, your understanding of the guard is very deep—I’ve learned a lot from you. Now, let's fix it according to your suggestions.

thenumberouscode commented:
@pytorchbot label "ciflow/trunk"

thenumberouscode commented:
@anijain2305 Hi, A workflow test has failed (https://github.com/pytorch/pytorch/actions/runs/16063185244/job/45383170740?pr=157566) due to a timeout, as indicated in the log. I suspect this issue is unrelated to my bug fix. Could we please retry the test? Thank you!

thenumberouscode commented:
> > @mlazos @anijain2305 I have added a cache for dynamically created user functions decorated with torch._dynamo.disable, allowing us to use the real user function IDs with the ID_MATCH guard without needing recompilation. Please check out my new code and share your feedback when you get a chance.
> 
> Can you explain a little more how it works? looking at the code it isn't clear to me why it works haha

Sure, the call of resume functions always have the following format: 
```__resume_at_xx(torch._dynamo.disable(_create_nested_fn(my_fn, '__co_consts[4]', 'my_fn', None, None, None, None)), x)```
 The first argument will always be a new object, so we cannot check its ID. Instead, we can check my_fn, which is the user's function. That's why I added caching in the _create_nested_fn function; this ensures that the user function's ID does not change. Additionally, in torch._dynamo.disable, the original user function is recorded in _torchdynamo_orig_callable.https://github.com/pytorch/pytorch/blob/6f23f53599629a47d6e097b2a027048658a142d4/torch/_dynamo/eval_frame.py#L1002

mlazos commented:
> > > @mlazos @anijain2305 I have added a cache for dynamically created user functions decorated with torch._dynamo.disable, allowing us to use the real user function IDs with the ID_MATCH guard without needing recompilation. Please check out my new code and share your feedback when you get a chance.
> > 
> > 
> > Can you explain a little more how it works? looking at the code it isn't clear to me why it works haha
> 
> Sure, the call of resume functions always have the following format: `__resume_at_xx(torch._dynamo.disable(_create_nested_fn(my_fn, '__co_consts[4]', 'my_fn', None, None, None, None)), x)` The first argument will always be a new object, so we cannot check its ID. Instead, we can check my_fn, which is the user's function. That's why I added caching in the _create_nested_fn function; this ensures that the user function's ID does not change. Additionally, in torch._dynamo.disable, the original user function is recorded in _torchdynamo_orig_callable.
> 
> https://github.com/pytorch/pytorch/blob/6f23f53599629a47d6e097b2a027048658a142d4/torch/_dynamo/eval_frame.py#L1002

Ah this makes sense now, I didn't realize how `_create_nested_fn` was used, but basically you now will just return the same object any time the same code is passed in. So one question I have is when will the cache ever be cleared? we should clear it once the output code using that entry is deleted. An example finalizer I added is [here](https://github.com/pytorch/pytorch/blob/e98dd95446e009ace1722498effbf32250d623e4/torch/_dynamo/variables/optimizer.py#L408C19-L408C38).

thenumberouscode commented:
> > > > @mlazos @anijain2305 I have added a cache for dynamically created user functions decorated with torch._dynamo.disable, allowing us to use the real user function IDs with the ID_MATCH guard without needing recompilation. Please check out my new code and share your feedback when you get a chance.
> > > 
> > > 
> > > Can you explain a little more how it works? looking at the code it isn't clear to me why it works haha
> > 
> > 
> > Sure, the call of resume functions always have the following format: `__resume_at_xx(torch._dynamo.disable(_create_nested_fn(my_fn, '__co_consts[4]', 'my_fn', None, None, None, None)), x)` The first argument will always be a new object, so we cannot check its ID. Instead, we can check my_fn, which is the user's function. That's why I added caching in the _create_nested_fn function; this ensures that the user function's ID does not change. Additionally, in torch._dynamo.disable, the original user function is recorded in _torchdynamo_orig_callable.
> > https://github.com/pytorch/pytorch/blob/6f23f53599629a47d6e097b2a027048658a142d4/torch/_dynamo/eval_frame.py#L1002
> 
> Ah this makes sense now, I didn't realize how `_create_nested_fn` was used, but basically you now will just return the same object any time the same code is passed in. So one question I have is when will the cache ever be cleared? we should clear it once the output code using that entry is deleted. An example finalizer I added is [here](https://github.com/pytorch/pytorch/blob/e98dd95446e009ace1722498effbf32250d623e4/torch/_dynamo/variables/optimizer.py#L408C19-L408C38).
> 
> We should also clear it if `torch._dynamo.reset()` is called. Otherwise this looks great!

After conducting a local test, I've discovered that the finalizer in OutputGraph is not functioning in this scenario. Since our cache is shared across different OutputGraph instances, we cannot clear it during the exit process of a single graph. Our cache needs to persist throughout the entire compilation process, similar to how code caches operate.

To address this issue, we should implement a CreateNestedFnCache to manage all function caches, which can then be cleared using torch._dynamo.reset().

Please take a look at my new commit to see the implementations, and don't hesitate to share any feedback or suggestions!

mlazos commented:
> Please take a look at my new commit to see the implementations, and don't hesitate to share any feedback or suggestions!

My thought here is that each output graph can add a finalizer to clear the entries from the cache that it added, not necessarily erase the entire cache. Lmk if this makes sense.

thenumberouscode commented:
> > Please take a look at my new commit to see the implementations, and don't hesitate to share any feedback or suggestions!
> 
> My thought here is that each output graph can add a finalizer to clear the entries from the cache that it added, not necessarily erase the entire cache. Lmk if this makes sense.

After a  long time reviewing the source code, I still believe that the cache for create_nested_fn cannot be cleared by the corresponding finalizer of OutputGraph; it should persist for as long as the code cache exists. I will illustrate this with an example from the issue reproduction.
```python
@torch.compile(backend="eager")
def fn(x):
    @torch._dynamo.disable
    def inner(x):
        return x + 10

    return inner(x)
```
After compilation, the equivalent Python code looks like this, simplified for easier understanding.
```python
def resume_in_fn_at_12(___stack0, x):
    inner = ___stack0
    return inner(x) + 1

def fn(x):
    return resume_in_fn_at_12(torch._dynamo.
        disable( _create_nested_fn(inner, '__co_consts[3]', 'inner', None, None, None,None)), x))
```
There will be two OutputGraph instances generating the code mentioned above. OutputGraph 1 is in charge of compiling the fn, while OutputGraph 2 takes care of compiling resume_in_fn_at_12. However, the line `torch._dynamo.disable(_create_nested_fn(inner, '__co_consts[3]', 'inner', None, None, None, None))` runs in the default Python interpreter because it breaks the graph, meaning there's no OutputGraph around to record its finalizer. Plus, if we clear the cache somehow, The next time fn is called, _create_nested_fn will always create a new function object for inner, giving it a new ID that breaks the ID_MATCH guard and triggers a recompilation, which doesn’t really make sense.

thenumberouscode commented:
> > > Please take a look at my new commit to see the implementations, and don't hesitate to share any feedback or suggestions!
> > 
> > 
> > My thought here is that each output graph can add a finalizer to clear the entries from the cache that it added, not necessarily erase the entire cache. Lmk if this makes sense.
> 
> After a long time reviewing the source code, I still believe that the cache for create_nested_fn cannot be cleared by the corresponding finalizer of OutputGraph; it should persist for as long as the code cache exists. I will illustrate this with an example from the issue reproduction.
> 
> ```python
> @torch.compile(backend="eager")
> def fn(x):
>     @torch._dynamo.disable
>     def inner(x):
>         return x + 10
> 
>     return inner(x)
> ```
> 
> After compilation, the equivalent Python code looks like this, simplified for easier understanding.
> 
> ```python
> def resume_in_fn_at_12(___stack0, x):
>     inner = ___stack0
>     return inner(x) + 1
> 
> def fn(x):
>     return resume_in_fn_at_12(torch._dynamo.
>         disable( _create_nested_fn(inner, '__co_consts[3]', 'inner', None, None, None,None)), x))
> ```
> 
> There will be two OutputGraph instances generating the code mentioned above. OutputGraph 1 is in charge of compiling the fn, while OutputGraph 2 takes care of compiling resume_in_fn_at_12. However, the line `torch._dynamo.disable(_create_nested_fn(inner, '__co_consts[3]', 'inner', None, None, None, None))` runs in the default Python interpreter because it breaks the graph, meaning there's no OutputGraph around to record its finalizer. Plus, if we clear the cache somehow, The next time fn is called, _create_nested_fn will always create a new function object for inner, giving it a new ID that breaks the ID_MATCH guard and triggers a recompilation, which doesn’t really make sense. cc @mlazos

@mlazos Please check the reply and let me know your feedback at your convenience.

mlazos commented:
Ah this makes sense, the caching occurs during during default python interpreter execution, so it isn't associated to any specific output graph artifact. This is okay with me.

anijain2305 commented:
@pytorchbot merge

mlazos commented:
> @anijain2305 @mlazos The workflow failed on the test_autograd.py::TestAutograd::test_saved_tensor_hooks_extra_enter_during_bw_no_leak. ([pytorch/pytorch/actions/runs/16361070389/job/46232942229?pr=157566](https://github.com/pytorch/pytorch/actions/runs/16361070389/job/46232942229?pr=157566)) The reason is that the test case is executed with **Dynamo wrapped,** the unpack function will hit a graph break. As a result, the unpack function gets cached, leading to a failure in the test case because the unpack_hook_ref value will not be None. I Just add @skipIfTorchDynamo on test_saved_tensor_hooks_extra_enter_during_bw_no_leak to solve it. Please let me know if this makes sense
> 
> test_saved_tensor_hooks_extra_enter_during_bw_no_leak:
> 
> ```python
> def test_saved_tensor_hooks_extra_enter_during_bw_no_leak(self):
>         # This usage of saved tensor is not supported, but should not leak
>         def scope():
>             def unpack(x):
>                 weak_ctx_1().__enter__()
>                 return x
> 
>             ctx_1 = torch.autograd.graph.saved_tensors_hooks(lambda x: x, unpack)
>             weak_ctx_1 = weakref.ref(ctx_1)
> 
>             x = torch.randn(3, 3, requires_grad=True)
>             with ctx_1:
>                 x.sin().sum().backward()
>             return weakref.ref(unpack)
> 
>         with disable_gc():
>             unpack_hook_ref = scope()
>             self.assertIsNone(unpack_hook_ref())
> ```
> 
> The error stack: <img alt="image" width="979" height="384" src="https://private-user-images.githubusercontent.com/24456053/467941674-dd387b77-1532-4b9e-a7e0-d21f278c7ce0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTMwNTIwNDIsIm5iZiI6MTc1MzA1MTc0MiwicGF0aCI6Ii8yNDQ1NjA1My80Njc5NDE2NzQtZGQzODdiNzctMTUzMi00YjllLWE3ZTAtZDIxZjI3OGM3Y2UwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzIwVDIyNDkwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdmNGZmMzU2NDk0MjM0OTk2NWE0NWYzYTMzMWY3YjgxZDNlYzc0MDhjYmEyOGRmMjQyNzYwYzZjMzIwZWM3OWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Js885HomhpcMqJjc8_UVhR3LhbtlskRUfbge0NfwQDw">

Could add a properly placed torch._dynamo.reset() instead of skipping it entirely? It would be good to test that the code isn't causing a leak beyond the way we expect.

thenumberouscode commented:
> > @anijain2305 @mlazos The workflow failed on the test_autograd.py::TestAutograd::test_saved_tensor_hooks_extra_enter_during_bw_no_leak. ([pytorch/pytorch/actions/runs/16361070389/job/46232942229?pr=157566](https://github.com/pytorch/pytorch/actions/runs/16361070389/job/46232942229?pr=157566)) The reason is that the test case is executed with **Dynamo wrapped,** the unpack function will hit a graph break. As a result, the unpack function gets cached, leading to a failure in the test case because the unpack_hook_ref value will not be None. I Just add @skipIfTorchDynamo on test_saved_tensor_hooks_extra_enter_during_bw_no_leak to solve it. Please let me know if this makes sense
> > test_saved_tensor_hooks_extra_enter_during_bw_no_leak:
> > ```python
> > def test_saved_tensor_hooks_extra_enter_during_bw_no_leak(self):
> >         # This usage of saved tensor is not supported, but should not leak
> >         def scope():
> >             def unpack(x):
> >                 weak_ctx_1().__enter__()
> >                 return x
> > 
> >             ctx_1 = torch.autograd.graph.saved_tensors_hooks(lambda x: x, unpack)
> >             weak_ctx_1 = weakref.ref(ctx_1)
> > 
> >             x = torch.randn(3, 3, requires_grad=True)
> >             with ctx_1:
> >                 x.sin().sum().backward()
> >             return weakref.ref(unpack)
> > 
> >         with disable_gc():
> >             unpack_hook_ref = scope()
> >             self.assertIsNone(unpack_hook_ref())
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The error stack: <img alt="image" width="979" height="384" src="https://private-user-images.githubusercontent.com/24456053/467941674-dd387b77-1532-4b9e-a7e0-d21f278c7ce0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTMwNTIwNDIsIm5iZiI6MTc1MzA1MTc0MiwicGF0aCI6Ii8yNDQ1NjA1My80Njc5NDE2NzQtZGQzODdiNzctMTUzMi00YjllLWE3ZTAtZDIxZjI3OGM3Y2UwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzIwVDIyNDkwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdmNGZmMzU2NDk0MjM0OTk2NWE0NWYzYTMzMWY3YjgxZDNlYzc0MDhjYmEyOGRmMjQyNzYwYzZjMzIwZWM3OWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Js885HomhpcMqJjc8_UVhR3LhbtlskRUfbge0NfwQDw">
> 
> Could add a properly placed torch._dynamo.reset() if torch._dynamo.is_compiling() instead of skipping it entirely? It would be good to test that the code isn't causing a leak beyond the way we expect.

@mlazos I’ve removed the skipIfDynamo decorator and added torch._dynamo.reset() in the unit test. Please review it again and leave any comments.

mlazos commented:
Looks good thanks!

thenumberouscode commented:
> Looks good thanks!

@mlazos The workflow failed again for a unrelated lint problem. Could we just run the workflow again? thanks.
the failed workflows:
1. https://github.com/pytorch/pytorch/actions/runs/16433277272/job/46683063908?pr=157566 (lint failure)
2. https://github.com/pytorch/pytorch/actions/runs/16433277255/job/46683401413?pr=157566 (unstable workflow)

thenumberouscode commented:
@mlazos @anijain2305 Can we trigger another workflow run? Appreciate it.

yangw-dev commented:
@pytorchbot  revert -m "failed an odd internal test, please reach out to metamate to fix it, D79112610" -c ghfirst

thenumberouscode commented:
@yangw-dev Can you let me know which unit test failed? cc @mlazos

mlazos commented:
@thenumberouscode can you rebase this PR and then I'll merge again? I don't see any failures.

mlazos commented:
oh @thenumberouscode you'll have to rebase manually w/ cherry-picks because the PR was already merged. Once that's done, push and I can merge

thenumberouscode commented:
> oh @thenumberouscode you'll have to rebase manually w/ cherry-picks because the PR was already merged and reverted. Once that's done, push and I can merge

@mlazos I've done the rebase. but there is no commit in this PR. Do we need to open a new PR using the current git branch to complete the merge ?

thenumberouscode commented:
@mlazos new PR is https://github.com/pytorch/pytorch/pull/160707. thx.
